hedgehog/src/Hedgehog/Internal/Property.hs

```diff
@@ -1021,6 +1048,58 @@ coverageFailures :: TestCount -> Coverage CoverCount -> [Label CoverCount]
 coverageFailures tests (Coverage kvs) =
   filter (not . labelCovered tests) (Map.elems kvs)
 
+-- | Is true when the test coverage satisfies the specified 'Confidence'
+--   contstraint for all 'Coverage CoverCount's
+confidenceSuccess :: TestCount -> Confidence -> Coverage CoverCount -> Bool
+confidenceSuccess tests confidence =
+  let
+    assertLow :: Label CoverCount -> Bool
+    assertLow MkLabel{..} =
+      wilsonLowerBound
+        (fromIntegral $ unCoverCount labelAnnotation)
+        (fromIntegral tests)
+        (1 / fromIntegral (unConfidence confidence))
+      >= 0.9 * (unCoverPercentage labelMinimum / 100.0)
+  in
+    and . fmap assertLow . Map.elems . coverageLabels
+
+-- | Is true when there exists a label that is sure to have failed according to
+--   the 'Confidence' constraint
+confidenceFailure :: TestCount -> Confidence -> Coverage CoverCount -> Bool
+confidenceFailure tests confidence =
+  let
+    assertHigh :: Label CoverCount -> Bool
+    assertHigh MkLabel{..} =
+      wilsonUpperBound
+        (fromIntegral $ unCoverCount labelAnnotation)
+        (fromIntegral tests)
+        (1 / fromIntegral (unConfidence confidence))
+      < (unCoverPercentage labelMinimum / 100.0)
+  in
+    or . fmap assertHigh . Map.elems . coverageLabels
+
+-- In order to get an accurate measurement with small sample sizes, we're
+-- using the Wilson score interval
+-- (<https://en.wikipedia.org/wiki/Binomial_proportion_confidence_interval#Wilson_score_interval
+-- wikipedia>) instead of a normal approximation interval.
+wilson :: Integer -> Integer -> Double -> Double
+wilson k n z =
+  let
+    nf = fromIntegral n
+    p = fromIntegral k / fromIntegral n
+  in
+    (p + z * z/(2 * nf) + z * sqrt (p * (1 - p) / nf + z * z / (4 * nf * nf)))
+    /
+    (1 + z * z / nf)
+
+wilsonLowerBound :: Integer -> Integer -> Double -> Double
+wilsonLowerBound k n a =
+  wilson k n $ invnormcdf (a / 2)
```

Comment:
I don't think this is the right calling convention for `wilson`. `z` is the probit for the error bounds, and is included in the denominator of wilson as `z^2`. I think you still need to use `1  - a / 2`, but with a negative sign, in `wilson`, not a plus.

This is throwing off the lower bound by quite a margin, especially with confidence intervals around 10^9.

Comparing with a beta distribution with alpha 5, beta 10, and 10^7, it's 0.045 when it should be more like 0.008.